File: src/app/model/BearingMetric.ts
import { LocationServiceEvent } from './LocationServiceEvent';
import { MetricService } from "../services/metric.service";
import { Metric, MetricType } from "./Metric";

export class BearingMetric extends Metric {
    private intervalId: any;

    constructor(metricService: MetricService) {
        super(MetricType.BEARING, 'Bearing', '°', metricService, '°');
    }

    startLogging(): void {
        if (this.intervalId) return;
        this.intervalId = setInterval(() => {
            this.bearingUpdateHandler();
        }, 1000); // Update every second
    }

    stopLogging(): void {
        if (this.intervalId) {
            clearInterval(this.intervalId);
            this.intervalId = null;
        }
    }

    bearingUpdateHandler(): void {
        const bearing = this.metricService.getLocationService().bearingForHorizontalPhone;
        this.addValue(bearing, new Date());
    }
}

File: android/app/src/main/java/com/openbikecomputer/model/BearingMetric.java
package com.openbikecomputer.model;

import com.openbikecomputer.service.IMetricService;

import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.TimeUnit;

public class BearingMetric extends Metric {
    private ScheduledExecutorService executorService;
    private boolean started = false;

    public BearingMetric(IMetricService metricService) {
        super(MetricType.BEARING, "Bearing", "\u00b0", metricService, "\u00b0");
    }

    @Override
    public void startLogging() {
        if (started) {
            return;
        }
        started = true;
        executorService = Executors.newSingleThreadScheduledExecutor(r -> {
            Thread thread = new Thread(r, "BearingMetric");
            thread.setDaemon(true);
            return thread;
        });
        executorService.scheduleAtFixedRate(this::bearingUpdateHandler, 1000, 1000, TimeUnit.MILLISECONDS);
    }

    @Override
    public void stopLogging() {
        if (executorService != null) {
            executorService.shutdownNow();
            executorService = null;
        }
        started = false;
    }

    public void bearingUpdateHandler() {
        double bearing = metricService.getLocationService().getBearingForHorizontalPhone();
        addValue(bearing, System.currentTimeMillis());
    }
}
